source/code/plugins/agent_telemetry_script.rb

```diff
@@ -0,0 +1,371 @@
+      File.open(@os_info).each_line do |line|
+        agent_telemetry.OSDistro = line.sub("OSName=","").strip if line =~ /OSName/
+        agent_telemetry.OSVersion = line.sub("OSVersion=","").strip if line =~ /OSVersion/
+      end
+      agent_telemetry.Region = OMS::Configuration.azure_region
+      agent_telemetry.ConfigMgrEnabled = File.exist?("/etc/opt/omi/conf/omsconfig/omshelper_disable").to_s
+      agent_telemetry.ResourceUsage = calculate_resource_usage
+      agent_telemetry.QoS = calculate_qos
+      log_info(agent_telemetry.inspect) if @verbose
+      return agent_telemetry
+    end
+
+    def heartbeat()
+      log_info("Agent Telemetry Script Heartbeat.")
+
+      # Reload config in case of updates since last topology request
+      @load_config_return_code = load_config
+      if @load_config_return_code != 0
+        log_error("Error loading configuration from #{@omsadmin_conf_path}")
+        return @load_config_return_code
+      end
+
+      # Check necessary inputs
+      if @workspace_id.nil? or @agent_guid.nil? or @url_tld.nil? or
+        @workspace_id.empty? or @agent_guid.empty? or @url_tld.empty?
+        log_error("Missing required field from configuration file: #{@omsadmin_conf_path}")
+        return OMS::MISSING_CONFIG
+      elsif !OMS::Common.file_exists_nonempty(@cert_path) or !OMS::Common.file_exists_nonempty(@key_path)
+        log_error("Certificates for topology request do not exist")
+        return OMS::MISSING_CERTS
+      end
+
+      # Generate the request body
+      body = create_body.to_json
+
+      # Form headers
+      headers = {}
+      req_date = Time.now.utc.strftime("%Y-%m-%dT%T.%N%:z")
+      headers[OMS::CaseSensitiveString.new("x-ms-Date")] = req_date
+      headers["User-Agent"] = "LinuxMonitoringAgent".concat(OMS::Common.get_agent_version)
+      headers[OMS::CaseSensitiveString.new("Accept-Language")] = "en-US"
+
+      # Form POST request and HTTP
+      uri = "https://#{@workspace_id}.oms.#{@url_tld}/AgentService.svc/AgentTelemetry"
+      req,http = OMS::Common.form_post_request_and_http(headers, uri, body,
+                      OpenSSL::X509::Certificate.new(File.open(@cert_path)),
+                      OpenSSL::PKey::RSA.new(File.open(@key_path)), @proxy_path)
+      
+      log_info("Generated telemetry request:\n#{req.body}") if @verbose
+
+      # Submit request
+      begin
+        res = nil
+        res = http.start { |http_each| http.request(req) }
+      rescue => e
+        log_error("Error sending the telemetry request to OMS agent management service: #{e.message}")
+      end
+
+      if !res.nil?
+        log_info("OMS agent management service topology request response code: #{res.code}") if @verbose
+      
+        if res.code == "200"
+          log_info("OMS agent management service topology request success")
+          return 0
+        else
+          log_error("Error sending OMS agent management service topology request. HTTP code #{res.code}")
```

Comment:
to fix: still getting a 400 response.

Reply:
Fixed, service side not expecting all the fields yet. Sending limited fields in the meantime...

Reply:
fixed: incorrect serialization of telemetry object